Add popup support for documentation and previews

nickspoons opened the pull request:
This PR adds support for the recently added Vim 8.1 popup API, and neovim's recently added floating windows. The end goal is to use popups instead of the preview window whereever possible, plus providing nicer documentation views with cleaner structure and some syntax highlighting.

Here's an example, Vim 8.1.2418 in WSL:
[![asciicast](https://asciinema.org/a/AgxMKKwBmhfSEcbS3y50lb34l.svg)](https://asciinema.org/a/AgxMKKwBmhfSEcbS3y50lb34l)

Vim also has the `popup` `completeopt` option which already works without this PR, displaying documentation in a popup when navigating through the completion menu:

```vim
set completeopt+=popup
" or
set completeopt+=popuphidden
```

However this popup by default contains only be a single string with a single colour. We can now replace this with a highlighted documentation window (as if the user had manually requested documentation). This solution uses a `GetDocumentation` request for properties and a `SignatureHelp` request for methods, attempting to display the best available documentation for different completion types.

[![asciicast](https://asciinema.org/a/8pEOWIvlkdRpV9JZCPNU4AraO.svg)](https://asciinema.org/a/8pEOWIvlkdRpV9JZCPNU4AraO)

Enable popups by setting the new variable:

```vim
" This variable is likely to change before this PR is merged
let g:OmniSharp_prefer_popups = 1

set completepopup=highlight:Pmenu,border:off
set completeopt+=popuphidden
```

In Vim, documentation is now displayed in popups, and `:OmniSharpPreviewDefinition` and `:OmniSharpPreviewImplementation` both use popups.

**Note:** These popup/floating window APIs are very new and require very recent versions of Vim/neovim

~There is currently **no** neovim support.~ Edit 19-03-20: Neovim support finally added.

**TODO**

* [X] Open documentation/signatureHelp in popup
* [X] Open definition/implementation/metadata preview in popup
* [x] Improve documentation structure and provide syntax highlighting
* [X] Full highlighted documentation popup for completion menu item documentation
* [x] Customisability
  * [x] Looks: options for borders, background colour etc.
  * [x] Position: options for where the popup is positioned (atcursor, fullscreen, peek)
  * [x] Interaction: keybindings to close and scroll popups
* [X] Add equivalent support for neovim windows
* [x] Make poups/floating windows default (opt-out) for versions of Vim/neovim that support them

**Followup**

_Completion popups_: Popups give us the potential to completely change the completion menus. Instead of a menu containing all overloads for a method in the completion list, we could just display a single one, and show the overloads in a popup? Or provide a popupmenu mapping to rotate through the signaturehelp docs for a completion item method and its arguments?

However there will be issues with how these systems work with snippets so it will take some thought and design.

_Selector UI_: We can add a new "Selector UI" option, providing a popup menu for e.g. CodeActions instead of FZF/CtrlP/Unite.

nickspoons commented:
I have extended the documentation (both in popups and in the preview window), it now includes parameter information, Exceptions, and return information when available. There is also some simple syntax highlighting to make it easier to parse.

SignatureHelp documentation hasn't been extended in the same way but it might be possible to create some new functionality to toggle through overloads? I'll have a think about it, that could be very useful.

jpfeiffer16 commented:
@nickspoons. This is great stuff. The preview popup looks awesome. 

I agree on the signature help stuff. Toggling though overloads would be great. The current implementation is somewhat annoying for me, but I haven't dug in to figure out how you would change it. I think it would be better to have a VS style popup that shows all parameters with the current one highlighted. Not sure how to go about doing it though.

nickspoons commented:
@jpfeiffer16 yeah I was thinking of something along those lines too. It could be displaying a full documentation window like the `:OmniSharpDocumentation` popup, with a text property highlighting the current argument in bold or italics or something - it would have to be customisable because bold/italics won't work in some terminals.

Then `<C-n>`/`<C-p>` (?) mappings to toggle through overloads would be cool. With something similar when viewing completions.

jpfeiffer16 commented:
Yea. Not sure how we would map the bindings but there's got to be a way to do it.

nickspoons commented:
@jpfeiffer16 I've improved the signature help documentation to display the signature, active parameter with parameter description, full signature summary and return information (I excluded the full param list with documentation as it becomes redundant, and also left exceptions out - they can be seen with the `:OmniSharpDocumentation` view).

I have also highlighted the active parameter in the first line (currently hard coded to highlight with bold,underline,italic), and show how many overloads there are.

Next, I'll add some (eventually customisable) key-bindings to navigate through the overloads, and possibly also to navigate left and right through the arguments.

jpfeiffer16 commented:
@nickspoons . Awesome! It looks great. I did notice the param docs didn't show up but I might have been on a type with no docs. I'll try to get some more testing on it later today.

nickspoons commented:
@jpfeiffer16 I've just added some filters, allowing you to scroll popups, and navigate signatureHelp popups.

The default and so-far hardcoded filters are:

- `<Esc>`: Close the popup (`j` and `k` also work)
- `<C-e>`: Scroll down 1 line
- `<C-y>`: Scroll up 1 line
- `<C-d>`: Scroll down half a page
- `<C-u>`: Scroll up half a page
- `<C-f>`: Scroll down 1 page
- `<C-b>`: Scroll up 1 page

Signature help:

- `<C-n>`: Next signature overload
- `<C-p>`: Previous signature overload
- `<C-h>`: Previous parameter
- `<C-l>`: Next parameter

I added an asciinema to the PR description to demonstrate. Popup positioning is an issue, as can be seen in the screencast.

jpfeiffer16 commented:
@nickspoons awesome!
I wasn't getting full signature help on the `string.Format()` like you were though. It seems like it sometimes works if you cycle through all of the overloads on the `string.Join()` for example. I wonder if it's framework specific. Are you on a mono project there?
The keybindings are amazing though and this experience is shaping up to be way better and more powerful than the vs popups ever were.

nickspoons commented:
@jpfeiffer16 yes it's a bit strange about that documentation, it's come up in a various issues: #508, #533 

The demo in the description is a native `dotnet` project, in an archlinux environment running in WSL, and I'm getting pretty good documentation since upgrading a few bits and pieces (described in #533). But I often get different documentation depending on which SDKs I have installed, and using .NET Framework vs. .NET Core.

At least the new popup keybindings allow us to easily see which overloads actually have docs! Very interesting that only _some_ of the `string.Join()` overloads had docs in the demo...

jpfeiffer16 commented:
@nickspoons ah.. Makes sense.
Already using the heck out of it!

nosami commented:
@nickspoons That looks amazing. Are you on twitter by the way? https://twitter.com/JasonImison/status/1202976554906730496?s=20

nickspoons commented:
Thanks @nosami - but no, I'm not on Twitter.

nickspoons commented:
"Rich" documentation popup added for completions - see the updated PR description for details and another screencast. Note that although this popup looks the same as the `SignatureHelp` popup, there are no navigation mappings enabled. We _may_ be able to enable them again once the configuration options are included, allowing users to set their own mappings, but it still may not work well, as conflicts with default popup-menu bindings (`<C-n>`/`<C-p>` etc.) _and_ user mappings begin to build up.

Since there currently is no way to toggle the documentation to next/previous parameter, it might be a good idea to remove the "active parameter" highlighting and parameter description. When users want more they can always cancel/accept the completion and trigger signature help again for the "full version".

Also, some work may need to be done on popup sizing/positioning. At this stage I'm unsure if it's a good idea to start manually handling this, or not.

nickspoons commented:
As mentioned in my previous comment, I have removed the active parameter highlighting from completion documentation, and instead included the complete Parameters and Exceptions documentation (as seen in `:OmniSharpDocumentation`).

nickspoons commented:
Sorry for the slow progress on this, it's been a very busy summer. Hopefully I'll be able to get back to it soon.

nickspoons commented:
OK, initial neovim support is finally included!

I have dropped the Vim popup "filters" and instead only use temporary mappings for closing the popup with `<Esc>`, as well as scrolling (`<C-e>`, `<C-y>`, `<C-d>`, `<C-u>`, `<C-f>`, `<C-b>`) and signature help navigation (`<C-j>`, `<C-k>`, `<C-h>`, `<C-l>` in both insert and normal modes). This actually works much more reliably than the filters, in Vim as well as neovim.

Still to do is neovim text-props for signature-help active parameter highlighting, mapping configuration, and window placement configuration. Hopefully without another 3-month delay 😊

nickspoons commented:
All popup mappings are now configurable, and the configuration is documented in the README and under `:h 'g:OmniSharp.popup.mappings'`. Also neovim namespace/highlighting has been used to highlight active signature help parameters so this now looks very similar to Vim (which uses text properties).

TODO:
* Configurable window sizing/positioning
* Display rich completion documentation in neovim, as currently done in Vim. neovim doesn not have `completeopt=popup` [yet](https://github.com/neovim/neovim/issues/10996) so this will either be left for neovim to implement, or I'll use the `CompleteDone` autocmd and position the window manually.

DasOhmoff commented:
Just want to say that this is one of the greatest features ever <3 <3 <3. Is there some kind of date we expect it to be released? Or is there some way to use this feature in the "beta" stage?

nickspoons commented:
@DasOhmoff I'm actually quite embarrassed about how long I've been working on this - lots of time restraints over the last few months.

The window layout stuff is the only thing remaining that really should be done in this PR - the neovim completion rich-documentation can be done afterwards. I'm really really hoping to find the time (half an hour here, half an hour there) over the next week to get it done.

In the meantime you can can definitely use it as it is. You can either clone the branch directly from my repo, or use follow the github "command line instructions" for checking out and testing a PR as a branch of your existing OmniSharp-vim plugin:

```sh
git checkout -b nickspoons-feature-popups master
git pull https://github.com/nickspoons/omnisharp-vim.git feature-popups
```

nickspoons commented:
OK I'm pretty happy with this PR now. I need to sleep so won't merge it right now, but if anyone wants to give it a whirl first and keep an eye out for bugs or problems, that'd be great.

jpfeiffer16 commented:
I'll check it out again in vim. I'll be able to give neovim a good run on it starting Tuesday. Don't hold up the merge for me though, it looks great so far!
